### ISSUE TITLE
FireStore get and batchget with mask

### ISSUE THREAD

**Issue by Thaina** [googleapis/google-cloud-dotnet#2048]
In the firestore low level API there was [mask functionality for filtering the returned document](https://cloud.google.com/firestore/docs/reference/rest/v1beta1/projects.databases.documents/batchGet#request-body) only for data that needed. It useful when we have large document but using specific api that only interest in some field

Could we also have it in high level API?

**Comment by jskeet** [comment]
Pinging @schmidt-sebastian who owns the overall high-level API experience.

**Comment by Thaina** [comment]
Query are using `Select` method to mask return data. Only need mask for `GetSnapShot` and `GetAllSnapshots`

Maybe we could allow create mask for each `DocumentReference` and let Get/GetAll read from the mask?

**Comment by jskeet** [comment]
Personally I wouldn't want that - a field mask doesn't feel like it's logically part of a `DocumentReference`. I would much prefer it to be a parameter.

**Comment by Thaina** [comment]
Not sure if possible, can the `GetAllSnapshots` function get multiple document, each of it has difference mask?

If it is I would suggest `GetAllSnapshots` has an overload

```C#
// default
public Task<IList<DocumentSnapshot>> GetAllSnapshotsAsync(IEnumerable<DocumentReference> documents,CancellationToken cancellationToken = default(CancellationToken));

// with mask
public Task<IList<DocumentSnapshot>> GetAllSnapshotsAsync(IEnumerable<Tuple<DocumentReference,Mask>> documents,CancellationToken cancellationToken = default(CancellationToken));

////

firestoreDB.GetAllSnapshotsAsync(docs.Select((doc) => (doc,GetMask(doc))));
```

**Comment by jskeet** [comment]
I wouldn't want to use `Tuple` here, I don't think. We could potentially create a new `MaskedDocumentReference` class. But I think we should wait until we know whether the team want to support this functionality at all before going too far down the road of designing what it would look like.

**Comment by Thaina** [comment]
I just don't want it to be a class if it contain only a mask and a documentreference

**Comment by jskeet** [comment]
That's exactly what a `Tuple<DocumentReference, Mask>` would be, but in a richer way - one that would support other operations, and have meaningful names.

**Comment by Thaina** [comment]
@jskeet My point is it should be `struct`

**Comment by jskeet** [comment]
But the `System.Tuple<...>` types (which you suggested before) are classes. We could use `ValueTuple`, but that requires an extra dependency. Extension methods aren't as discoverable as regular methods, and can be overly broad. (You might want a tuple with those values for other reasons, and not want the extension methods.) In my experience a custom type designed specifically for the job is *much* more usable.

Note that the default value of such a struct (whether custom or `ValueTuple`) would be unusable though (as it would have two null references) - that wouldn't be ideal.

Anyway, as I said before, there's not much point in bike-shedding the design when we don't know whether it's even going to happen or not.

**Comment by Thaina** [comment]
@schmidt-sebastian Hello, how this feature was discussed. Is it already accepted or rejected?

**Comment by schmidt-sebastian** [comment]
I am generally in favor of this, but haven't been able to come up with a convincing API design. I'll prioritize this for August.

**Comment by Thaina** [comment]
Thank you very much

**Comment by Thaina** [comment]
Sadly the august was passed

**Comment by Thaina** [comment]
@schmidt-sebastian Are there any problem on this specific request?

**Comment by jskeet** [comment]
Sebastian is out at the moment. I'll ask another team member whether they can share the result of the discussion.

**Comment by Thaina** [comment]
@jskeet What happen? Is him alright?

**Comment by jskeet** [comment]
@Thaina: Yes, he's fine - but I'm not going to go into the personal lives of Googlers on a public web site.

**Comment by Thaina** [comment]
@jskeet That's OK I just worried

**Comment by jskeet** [comment]
Available in version 1.0.0-beta14.
